File: POS/C# 5/RoboterWPF/Models/LevelParser.cs
using System.Xml.Linq;

namespace RoboterWPF.Models;

public class LevelParser
{
    public int Width { get; private set; } = 10;
    public int Height { get; private set; } = 10;

    public List<RobotElement> ParseLevel(string filename)
    {
        var elements = new List<RobotElement>();

        try
        {
            var doc = XDocument.Load(filename);
            var root = doc.Root;

            if (root == null)
                return elements;

            // Parse Width and Height
            var widthElement = root.Element("Width");
            var heightElement = root.Element("Height");

            if (widthElement != null)
                Width = int.Parse(widthElement.Value);

            if (heightElement != null)
                Height = int.Parse(heightElement.Value);

            // Parse Fields
            var fieldsElement = root.Element("Fields");
            if (fieldsElement != null)
            {
                // Handle both "Field" and "XML_Cell" element names
                foreach (var field in fieldsElement.Elements())
                {
                    if (field.Name != "Field" && field.Name != "XML_Cell")
                        continue;

                    var element = new RobotElement();

                    var xElement = field.Element("X");
                    var yElement = field.Element("Y");
                    var typeElement = field.Element("Type");

                    if (xElement != null)
                        element.X = int.Parse(xElement.Value);

                    if (yElement != null)
                        element.Y = int.Parse(yElement.Value);

                    if (typeElement != null)
                    {
                        string typeText = typeElement.Value.ToLower();

                        if (typeText == "stone")
                        {
                            element.Type = ElementType.Obstacle;
                        }
                        else if (typeText == "robot")
                        {
                            element.Type = ElementType.Robot;
                        }
                        else if (typeText == "wall")
                        {
                            element.Type = ElementType.Wall;
                        }
                        else if (IsLetter(typeText))
                        {
                            element.Type = ElementType.Letter;
                            element.Letter = typeElement.Value.ToUpper();
                        }
                        else
                        {
                            // Default to wall for unknown types
                            element.Type = ElementType.Wall;
                        }
                    }

                    elements.Add(element);
                }
            }
        }
        catch (Exception ex)
        {
            System.Diagnostics.Debug.WriteLine($"Error parsing level: {ex.Message}");
        }

        return elements;
    }

    private static bool IsLetter(string str)
    {
        return str.Length == 1 && char.IsLetter(str[0]);
    }
}

File: POS/C++ 5/Roboter/Roboter/XML/level_parser.cpp
#include "level_parser.h"

#include <iostream>

static bool isLetter(const std::string& str) {
	return str.length() == 1 && std::isalpha(static_cast<unsigned char>(str[0]));
}

QList<RobotElement*> level_parser::parse_level(std::string filename, RobotMap* map)
{
	QList<RobotElement*> elements;
	tinyxml2::XMLDocument doc;
	if (doc.LoadFile(filename.c_str()) != tinyxml2::XML_SUCCESS) {
		return elements;
	}
	tinyxml2::XMLElement* root = doc.RootElement();
	if (!root) {
		return elements;
	}

	for (tinyxml2::XMLElement* elem = root->FirstChildElement(); 
		elem != nullptr; 
		elem = elem->NextSiblingElement()) {
		std::string name = elem->Name();

		if (name == "Width") {
			int width = elem->IntText(10);
			elem = elem->NextSiblingElement();

			int height = elem->IntText(10);
			map->setSize(width, height);
		}

		if (name == "Fields") {
			for (tinyxml2::XMLElement* field = elem->FirstChildElement();
				field != nullptr;
				field = field->NextSiblingElement()) 
			{
				RobotElement* robot_element = new RobotElement();
				robot_element->x = field->FirstChildElement("X")->IntText();
				robot_element->y = field->FirstChildElement("Y")->IntText();
				std::string type = field->FirstChildElement("Type")->GetText();
				std::transform(type.begin(), type.end(), type.begin(),
					[](unsigned char c) { return std::tolower(c); });

				if (type == "stone") {
					robot_element->type = RobotElement::Obstacle;
				}
				else if (type == "robot") {
					robot_element->type = RobotElement::Robot;
				}
				else if (isLetter(type)) {
					robot_element->type = RobotElement::Letter;
					robot_element->letter = QString::fromStdString(field->FirstChildElement("Type")->GetText());
				}
				elements.append(robot_element);
			}
		}
	}

	return elements;
}
